#Importing the walmart csv file from my google drive link

In [1]:
#Importing all dependencies
import pandas as pd
print(pd.__version__)
df= pd.read_csv('Walmart.csv', encoding_errors='ignore')
print(type(df))
#encoding errors would ignore any encoding errors that might arise while reading the csv file.


2.2.2
<class 'pandas.core.frame.DataFrame'>


In [2]:
#displaying the number of rows and columns in the file
print(df.shape)

#displaying the first five rows of the dataset
print(df.head())

print('\n\n')
#description of the entire data
print(df.describe())

(10051, 11)
   invoice_id   Branch         City                category unit_price  \
0           1  WALM003  San Antonio       Health and beauty     $74.69   
1           2  WALM048    Harlingen  Electronic accessories     $15.28   
2           3  WALM067  Haltom City      Home and lifestyle     $46.33   
3           4  WALM064      Bedford       Health and beauty     $58.22   
4           5  WALM013       Irving       Sports and travel     $86.31   

   quantity        date      time payment_method  rating  profit_margin  
0       7.0  05-01-2019  13:08:00        Ewallet     9.1           0.48  
1       5.0  08-03-2019  10:29:00           Cash     9.6           0.48  
2       7.0  03-03-2019  13:23:00    Credit card     7.4           0.33  
3       8.0  27-01-2019  20:33:00        Ewallet     8.4           0.33  
4       7.0  08-02-2019  10:37:00        Ewallet     5.3           0.48  



         invoice_id      quantity        rating  profit_margin
count  10051.000000  10020.000000

Now there are a couple of things to be taken into consideration. When we use describe method we can see that the column quantity doesn't have 10051 values , that means there are some missing values.

Next, unit price is not a numerical column and hence it is not included in the list of columns in describe function output.

In [3]:
#checking the data type of columns in the dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10051 entries, 0 to 10050
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   invoice_id      10051 non-null  int64  
 1   Branch          10051 non-null  object 
 2   City            10051 non-null  object 
 3   category        10051 non-null  object 
 4   unit_price      10020 non-null  object 
 5   quantity        10020 non-null  float64
 6   date            10051 non-null  object 
 7   time            10051 non-null  object 
 8   payment_method  10051 non-null  object 
 9   rating          10051 non-null  float64
 10  profit_margin   10051 non-null  float64
dtypes: float64(3), int64(1), object(7)
memory usage: 863.9+ KB


Here , we can observe that the unit price and quantity columns have missing data in the csv file. Also unit price is an object and not an integer which would make it difficult for us to perform operations on the unit price data.

In [4]:
#checking for duplicate records in the file
df.duplicated() #will return boolean value for each row whether it is duplicated or not

#finding the total sum of duplicated records
print(df.duplicated().sum())

51


Here, we have found out that we have duplicate values in our file

In [5]:
#checking for missing values in the file
df.isnull().sum()

invoice_id         0
Branch             0
City               0
category           0
unit_price        31
quantity          31
date               0
time               0
payment_method     0
rating             0
profit_margin      0
dtype: int64

In [6]:
#removing all the duplicate records in the file
df.drop_duplicates(inplace=True)
#inplace=True will make sure that the new records wherein all the duplicate values have been removed will be stored in the same dataframe itself and won't return a new object

#check for duplicates once again
print(df.duplicated().sum())

#checking the shape of the file
print(df.shape)
#Here the shape(specifically the rows) got reduced because the duplicate records were all removed

0
(10000, 11)


Now we have solved the duplicating issue by removing them. Now we have to deal with null values. We can see that there are 31 records where quantity and unit price have null values. Since it is difficult for us to understand the quantity and unit price we will drop those records

In [7]:
#dropping the records with null values (quantity and unit price)
df.dropna(inplace=True)
#inplace=True will make sure that the new records wherein all the null values have been removed will be stored in the same dataframe itself and won't return a new object

#checking if there are any more null values in the file
print(df.isnull().sum())

#checking the shape of the file after the null values are removed
print(df.shape)
#The shape(especially the rows) would be reduced even further since the null records have been dropped from the data frame

invoice_id        0
Branch            0
City              0
category          0
unit_price        0
quantity          0
date              0
time              0
payment_method    0
rating            0
profit_margin     0
dtype: int64
(9969, 11)


#Converting the string datatype of unit price into float for analysis purposes


Here, we can observe that the values for unit price has dollar in them. So it is not easy to convert the unit price directly to float values. For this reason, we will first have to replace the dollar sign with empty string and then carry out the typecasting to float.

In [8]:
#replacing the $ sign and then coverting it into float
df['unit_price']=df['unit_price'].str.replace('$','').astype('float')

In [10]:
#checking whether the typecast changes are visible in the dataframe
print(df.head())
print(df.info())

   invoice_id   Branch         City                category  unit_price  \
0           1  WALM003  San Antonio       Health and beauty       74.69   
1           2  WALM048    Harlingen  Electronic accessories       15.28   
2           3  WALM067  Haltom City      Home and lifestyle       46.33   
3           4  WALM064      Bedford       Health and beauty       58.22   
4           5  WALM013       Irving       Sports and travel       86.31   

   quantity        date      time payment_method  rating  profit_margin  
0       7.0  05-01-2019  13:08:00        Ewallet     9.1           0.48  
1       5.0  08-03-2019  10:29:00           Cash     9.6           0.48  
2       7.0  03-03-2019  13:23:00    Credit card     7.4           0.33  
3       8.0  27-01-2019  20:33:00        Ewallet     8.4           0.33  
4       7.0  08-02-2019  10:37:00        Ewallet     5.3           0.48  
<class 'pandas.core.frame.DataFrame'>
Index: 9969 entries, 0 to 9999
Data columns (total 11 columns):
 # 

#Addition of a new column

Now in our dataset we don't have a specific column indicating the total price or total amount for the number of quantities that each category has. As of now we only have the unit price and the quantities. So for that purpose we will create a new column called total amount.

In [11]:
#Checking and creating the new column based on unit price and quantity
print(df.columns)
df['total']= df['unit_price'] * df['quantity']
print(df.head())

Index(['invoice_id', 'Branch', 'City', 'category', 'unit_price', 'quantity',
       'date', 'time', 'payment_method', 'rating', 'profit_margin'],
      dtype='object')
   invoice_id   Branch         City                category  unit_price  \
0           1  WALM003  San Antonio       Health and beauty       74.69   
1           2  WALM048    Harlingen  Electronic accessories       15.28   
2           3  WALM067  Haltom City      Home and lifestyle       46.33   
3           4  WALM064      Bedford       Health and beauty       58.22   
4           5  WALM013       Irving       Sports and travel       86.31   

   quantity        date      time payment_method  rating  profit_margin  \
0       7.0  05-01-2019  13:08:00        Ewallet     9.1           0.48   
1       5.0  08-03-2019  10:29:00           Cash     9.6           0.48   
2       7.0  03-03-2019  13:23:00    Credit card     7.4           0.33   
3       8.0  27-01-2019  20:33:00        Ewallet     8.4           0.33   
4     

#Connecting the python to mysql

Now we want to carry out our further analysis through mysql and for that we need to connect our python to mysql. We need to install python libraries like pymysql and sqlalchemy for this purpose.

In [12]:
#installing pymysql and sqlalchemy packages

%pip install pymysql
%pip install sqlalchemy


   ---------------------------------------- 0.0/45.0 kB ? eta -:--:--
   ---------------------------------------- 0.0/45.0 kB ? eta -:--:--
   --------------------------- ------------ 30.7/45.0 kB ? eta -:--:--
   ---------------------------------------- 45.0/45.0 kB 445.3 kB/s eta 0:00:00
Note: you may need to restart the kernel to use updated packages.



In [13]:
#checking whether both the packages are correctly installed
%pip show pymysql
%pip show sqlalchemy

Name: PyMySQL
Version: 1.1.1
Summary: Pure Python MySQL Driver
Home-page: 
Author: 
Author-email: Inada Naoki <songofacandy@gmail.com>, Yutaka Matsubara <yutaka.matsubara@gmail.com>
License: MIT License
Location: c:\Users\Admin\anaconda3\Lib\site-packages
Requires: 
Required-by: 
Note: you may need to restart the kernel to use updated packages.
Name: SQLAlchemy
Version: 2.0.30
Summary: Database Abstraction Library
Home-page: https://www.sqlalchemy.org
Author: Mike Bayer
Author-email: mike_mp@zzzcomputing.com
License: MIT
Location: c:\Users\Admin\anaconda3\Lib\site-packages
Requires: greenlet, typing-extensions
Required-by: 
Note: you may need to restart the kernel to use updated packages.


In [14]:
#import dependencies for pymysql and sqlalchemy
import pymysql #this will work as an adapter, connecting our python with mysql
from sqlalchemy import create_engine #this also establishes a connection with sql, but this gives us provision to work with python objects rather than raw sql queries

Now before establishing a connection between python and mysql, we need to know a few things about our mysql connection



*   host=localhost (Since the mysql server is on my local machine itself)
*   port= 3306 (the port where the mysql server is running)
*   user = root (username of my mysql server)
*   password = admin (password of my mysql server)



In [15]:
help(df.to_sql)

Help on method to_sql in module pandas.core.generic:

to_sql(name: 'str', con, *, schema: 'str | None' = None, if_exists: "Literal['fail', 'replace', 'append']" = 'fail', index: 'bool_t' = True, index_label: 'IndexLabel | None' = None, chunksize: 'int | None' = None, dtype: 'DtypeArg | None' = None, method: "Literal['multi'] | Callable | None" = None) -> 'int | None' method of pandas.core.frame.DataFrame instance
    Write records stored in a DataFrame to a SQL database.

    Databases supported by SQLAlchemy [1]_ are supported. Tables can be
    newly created, appended to, or overwritten.

    Parameters
    ----------
    name : str
        Name of SQL table.
    con : sqlalchemy.engine.(Engine or Connection) or sqlite3.Connection
        Using SQLAlchemy makes it possible to use any DB supported by that
        library. Legacy support is provided for sqlite3.Connection objects. The user
        is responsible for engine disposal and connection closure for the SQLAlchemy
        conn

In [16]:
import pymysql
from sqlalchemy import create_engine

     # Update connection string with your correct credentials
engine_mysql = create_engine('mysql+pymysql://root:admin@localhost:3306/walmart_db')
# engine = create_engine("mysql+pymysql://user:pw@host/db", pool_pre_ping=True)

try:
    engine_mysql
    print('Connection Successful')
except:
         print('Connection Failed')



Connection Successful


After the connection is successful, we have to export the walmart dataframe to our sql database walmart_db inside our mysql. The engine is already established through create_engine and now we have to export the dataframe using that engine_mysql.

In [17]:
#exporting the dataframe to sql database
df.to_sql(name='walmart', con=engine_mysql, if_exists='append', index=False)

#name suggests the name of the table that would be created in the mysql database.
#if_exists suggests what action to be taken if the table already exists. In this case, the table data will be appended if the table already exists.
#Index refers to the index number 1,2,3,4 that is present in the dataframe. In this case, we don't want to export the index along with other details of the dataframe.

9969

In [18]:
df.shape

(9969, 12)